src/coprocessor/dag/rpn_expr/mod.rs

```diff
@@ -56,31 +60,15 @@ fn compare_mapper<F: CmpOp>(lhs_is_unsigned: bool, rhs_is_unsigned: bool) -> Box
     }
 }
 
-fn plus_mapper(lhs_is_unsigned: bool, rhs_is_unsigned: bool) -> Box<dyn RpnFunction> {
-    match (lhs_is_unsigned, rhs_is_unsigned) {
-        (false, false) => Box::new(RpnFnArithmetic::<IntIntPlus>::new()),
-        (false, true) => Box::new(RpnFnArithmetic::<IntUintPlus>::new()),
-        (true, false) => Box::new(RpnFnArithmetic::<UintIntPlus>::new()),
-        (true, true) => Box::new(RpnFnArithmetic::<UintUintPlus>::new()),
-    }
-}
-
-fn minus_mapper(lhs_is_unsigned: bool, rhs_is_unsigned: bool) -> Box<dyn RpnFunction> {
-    match (lhs_is_unsigned, rhs_is_unsigned) {
-        (false, false) => Box::new(RpnFnArithmetic::<IntIntMinus>::new()),
-        (false, true) => Box::new(RpnFnArithmetic::<IntUintMinus>::new()),
-        (true, false) => Box::new(RpnFnArithmetic::<UintIntMinus>::new()),
-        (true, true) => Box::new(RpnFnArithmetic::<UintUintMinus>::new()),
-    }
-}
-
-fn mod_mapper(lhs_is_unsigned: bool, rhs_is_unsigned: bool) -> Box<dyn RpnFunction> {
-    match (lhs_is_unsigned, rhs_is_unsigned) {
-        (false, false) => Box::new(RpnFnArithmetic::<IntIntMod>::new()),
-        (false, true) => Box::new(RpnFnArithmetic::<IntUintMod>::new()),
-        (true, false) => Box::new(RpnFnArithmetic::<UintIntMod>::new()),
-        (true, true) => Box::new(RpnFnArithmetic::<UintUintMod>::new()),
-    }
+macro_rules! int_arith {
+    ($Op:ident) => {
+        |lhs_is_unsigned, rhs_is_unsigned| match (lhs_is_unsigned, rhs_is_unsigned) {
+            (false, false) => Box::new($Op::<Int, Int>::func()),
```

Comment:
Maybe you can discover some ways to use generics instead of macros here :)

Reply:
With current design in arithmetic, I feel it's hard to use generic here... Plus<Int, Int> instead of Plus is a type, we cannot pass partial type, right?
The approach using generics I can think of is: create Plus, Minus... trait with associated types of parameters and return tupe, then implement them for tuples like (Int, Int). I think code will be much longer...

Reply:
I do want to use associated types. What stops me is the bug of default associated type (of course it's unstable yet). So we cannot write only the LHS type and let the other compiler make RHS and return type same as LHS by default...But writing all these three types every time we impl a trait is really annoying 😞